### PR TITLE
Create a `Results` class for storing metric output with associated metadata

### LINKED ISSUE #120: Feature: Create Result class to store all info about pipeline and results
In order to implement #95, we need to have access to all info about matcher params and metric params along with the eventual metric output and any other metadata needed to have a complete record of what was done. 

We should implement a Result object that is returned from `Metric.compute()` with the follwing info:

- Matcher info: a dictionary containing the matcher name and any parameters, e.g. `{”name”: IOU Matcher, “threshold”: 0.5}`
- Metrics info with name and parameters, similar to matcher
- Results (generic dictionary with whatever the metric wants)
- Traccuracy version (tag and commit)
- Other metadata to help users identify which run their results are from, including the "name" of the solution and gt TrackingGraph (this would require also letting users provide a name for their TrackingGraph), or date/time of run

We need the Matcher info to also be stored in the Matched object for continuity of information.
Then #95 is simply a to_dict and/or to_file function in this class. `run_metrics` can call that automatically with a user-provided or default location with one metric for file or all metrics in one file. Details TBD 🙂

### PR DESCRIPTION
**Pull request by msschwartz21**
# Proposed Change
Closes #120 

# Types of Changes
What types of changes does your code introduce? Put an x in the boxes that apply.
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature or enhancement
- [ ] Documentation update
- [ ] Tests and benchmarks
- [ ] Maintenance (e.g. dependencies, CI, releases, etc.)

Which topics does your change affect? Put an x in the boxes that apply.
- [ ] Loaders
- [x] Matchers
- [ ] Track Errors
- [x] Metrics
- [x] Core functionality (e.g. `TrackingGraph`, `run_metrics`, `cli`, etc.)

# Checklist
Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.

- [x] I have read the developer/contributing docs.
- [x] I have added tests that prove that my feature works in various situations or tests the bugfix (if appropriate).
- [x] I have checked that I maintained or improved code coverage.
- [x] I have checked the benchmarking action to verify that my changes did not adversely affect performance.
- [x] I have written docstrings and checked that they render correctly in the Read The Docs build (created after the PR is opened).
- [x] I have updated the general documentation including Metric descriptions and example notebooks if necessary.

### CODE REVIEW

File: src/traccuracy/metrics/_base.py [hunk 1]
```diff
@@ -21,13 +22,95 @@ class Metric(ABC):
...
+        matcher_info (optional, dict): Dictionary with matcher name and parameters
+        metric_info (optional, dict): Dictionary with metric name and parameters
+        gt_name (optional, str): Name of the ground truth data
+        pred_name (optional, str): Name of the predicted data
+    """
+
+    def __init__(
+        self,
+        results: dict | None = None,
+        matcher_info: dict | None = None,
```
Comment (cmalinmayor) [inline src/traccuracy/metrics/_base.py:82]: Results, matcher_info, and metric_info can be required and always included in the to_dict

File: tests/matchers/test_base.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+def test_matched_info():
+    matcher = DummyMatcher()
+    # Check that matcher info is correctly generated
+    assert matcher.info["name"] == "DummyMatcher"
+    # Only one key because there are no parameters
+    assert len(matcher.info.keys()) == 1
+
+    matcher = DummyMatcherParam()
+    # Check for parameter key
+    assert "param" in matcher.info
```
Comment (cmalinmayor) [inline tests/matchers/test_base.py:28]: Also test that the default value is present? And/or that if you initialize with not the default, it changes?

File: tests/test_run_metrics.py [hunk 3]
```diff
@@ -68,3 +68,4 @@ def test_run_metrics():
     assert results[0]["metric"].get("param") == "param1"
     assert results[1]["metric"]["name"] == "DummyMetricParam"
     assert results[1]["metric"].get("param") == "param2"
+    print(results)
```
Comment (cmalinmayor) [inline tests/test_run_metrics.py:71]: remove debugging print!

**Review by cmalinmayor — CHANGES_REQUESTED** [review]

**Review by cmalinmayor — APPROVED** [review]
Perfect! Thanks Morgan 🎊

**Comment by cmalinmayor** [comment]
Wait except all the tests failing @msschwartz21 😬

**Comment by msschwartz21** [comment]
> Wait except all the tests failing @msschwartz21 😬

I was being spacey... Should be ok now 🤞

**Comment by cmalinmayor** [comment]
@bentaculum Just warning you about these changes because I know you have built code on top of traccuracy output, and we plan to wrap that results dictionary in a class, so this will be a breaking change for you I suspect.

**Pull request merged**